infra/scripts/release-patch.sh

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env bash
+set -eo pipefail
+
+
+usage()
+{
+    echo "usage: release-patch.sh
+
+    This script is used to release a patch release. It is untested on major/minor releases and for those, some modification may be necessary.
+
+    -v, --version           version to release, example: 0.10.6
+    -t, --github-token      personal GitHub token
+    -c, --starting-commit   git commit hash; the script will cherry-pick that commit and all on master after it
+"
+}
+
+while [ "$1" != "" ]; do
+  case "$1" in
+      -v | --version )           VERSION="$2";            shift;;
+      -t | --github-token )      GH_TOKEN="$2";           shift;;
+      -c | --starting-commit )   STARTING_COMMIT="$2";    shift;;
+      -h | --help )              usage;                   exit;; 
+      * )                        usage;                   exit 1
+  esac
+  shift
+done
+
+if [ -z $VERSION ]; then usage; exit 1; fi
+if [ -z $GH_TOKEN ]; then usage; exit 1; fi
+regex="([0-9]+)\.([0-9]+)\.([0-9]+)"
+if [[ $VERSION =~ $regex ]]
+then
+    MAJOR="${BASH_REMATCH[1]}"
+    MINOR="${BASH_REMATCH[2]}"
+    PATCH="${BASH_REMATCH[3]}"
+else
+    usage
+    exit 1
+fi
+
+echo "This script is mostly idempotent; check git status for temp files before restarting. It will always prompt you before making any non-local change."
+
+# Go to infra/scripts directory
+cd $(dirname "$0")
+# Login to GitHub CLI
+echo $GH_TOKEN | gh auth login --with-token
+
+echo "Step 1: Cherry-pick new commits onto release branch"
+git fetch origin
+git checkout master
+last_commit=$(git rev-parse HEAD)
+git checkout v$MAJOR.$MINOR-branch
+cherrypick_from_master()
+{
+    echo "Cherrypicking commits"
+    git cherry-pick $STARTING_COMMIT^..$last_commit
+    echo "Commits cherrypicked"
+}
+if [ -z $STARTING_COMMIT ]; then
+    read -p "No starting commit given. Skip this step (y) or exit (n)? " choice
+    case "$choice" in
+        y|Y ) echo "Skipping this step" ;;
+        n|N ) exit ;;
+        * ) exit ;;
+    esac ;
+else
+    read -p "Will cherry-pick starting from commit $STARTING_COMMIT. Continue (y) or skip this step (n)? " choice
+    case "$choice" in
+        y|Y ) cherrypick_from_master ;;
+        n|N ) echo "Skipping this step" ;;
+        * ) echo "Skipping this step" ;;
+    esac ;
+fi
+
+commit_changelog()
+{
+    echo "Committing CHANGELOG.md"
+    git add ../../CHANGELOG.md
+    git commit -m "Update CHANGELOG for Feast v$MAJOR.$MINOR.$PATCH"
+}
+update_changelog()
+{
+    echo "Running changelog generator (will take up to a few minutes)"
+    echo -e "# Changelog\n" > temp \
+    && docker run -it --rm ferrarimarco/github-changelog-generator \
+    --user feast-dev \
+    --project feast  \
+    --release-branch master  \
```

Comment:
What about patch commits where the release branch is `v0.10-branch` and the release is `v0.10.4` for example?

Reply:
I think for patch commits both ways will work. I just did it off master since it was easier to test without starting an actual release (since if the release-branch is v0.10-branch I'd have to cherry-pick and push the v0.10.4 commits onto that branch first before being able to generate a changelog). There's also the escape hatch of inspecting and modifying the generated changelog but if basing off master ends up too error-prone we should change it to the branch

Reply:
That all makes sense, but its not clear to me how you ensure that the right changes make it into the changelog. How does the changelog generator know only to include unreleased commits on the v0.10-branch?

Reply:
My understanding was that most of the time we'd be adding all unreleased commits into the next patch, in which case I think these would be the same thing but correct me if I'm wrong here. Since the script prompts you to inspect and modify the generated changelog, the worst case scenario is just having to modify the file before continuing with the script

Reply:
We should only be backporting patches to a branch, but right now we are trying to release functionality a bit faster so we are even backporting functionality like feature inferencing. So by right we will rarely release all unreleased commits to branches. Otherwise there really isnt a point in maintaining the branches, we can just use master for releasing all commits.

So my preference here would be to ask the user which branch they want to release from, but I dont feel super strongly. Its also fair to just ask them to edit the file for now, but its more error prone I think.